File: gui/framecapture_gui/ipc/framing.py
"""The control channel's wire format (SPEC.md §15.1).

    4-byte little-endian length prefix + UTF-8 JSON body. Max message 64 KB.

The Python mirror of ``engine/core/ipc/framing.h``. Two implementations of one wire
format is a cost worth naming: they can drift. What makes it acceptable is that the
format is four lines of arithmetic, both sides are tested against the *same* stated
rules, and the alternative -- a C extension binding the C++ one -- would put a build
step between the GUI and its ability to start.

The prefix is the authority, not the pipe's message mode. The engine creates the pipe
with ``PIPE_TYPE_MESSAGE``, but this client reads it as a byte stream, so message
boundaries here come from the prefix alone. That is the arrangement ``IPC_PROTOCOL.md``
describes, and it is why the prefix exists at all.
"""

from __future__ import annotations

import json
from typing import Any

#: SPEC.md §15.1's ceiling. Applies to the JSON body; the prefix is on top.
MAX_MESSAGE_BYTES = 64 * 1024

#: Width of the length prefix, little-endian per §15.1.
LENGTH_PREFIX_BYTES = 4


class FramingError(Exception):
    """A frame that cannot be recovered from.

    Distinct from "not enough bytes yet", which is not an error and is signalled by
    returning ``None``. Conflating the two would make every partial read on a stream
    look like a failure.
    """


def encode_frame(body: str) -> bytes:
    """Prepend the length prefix to ``body``.

    Raises ``FramingError`` rather than emitting a frame the peer is obliged to reject:
    the sender is the side that can still do something about it.
    """
    encoded = body.encode("utf-8")
    if len(encoded) > MAX_MESSAGE_BYTES:
        raise FramingError(f"message of {len(encoded)} bytes exceeds the {MAX_MESSAGE_BYTES} byte limit")
    return len(encoded).to_bytes(LENGTH_PREFIX_BYTES, "little") + encoded


def encode_message(message: dict[str, Any]) -> bytes:
    """Serialise and frame one message."""
    # `separators` drops the whitespace `json.dumps` inserts by default. It is not
    # about bandwidth -- it is that the 64 KB ceiling should be spent on content.
    return encode_frame(json.dumps(message, separators=(",", ":")))


class FrameReader:
    """Reassembles frames from a byte stream, holding whatever is left over.

    One ``feed`` can yield several frames or none. Call ``next_frame`` until it returns
    ``None``.
    """

    def __init__(self) -> None:
        self._buffer = bytearray()

    def feed(self, chunk: bytes) -> None:
        self._buffer.extend(chunk)

    def next_frame(self) -> str | None:
        """The next complete frame, or ``None`` when more bytes are needed."""
        if len(self._buffer) < LENGTH_PREFIX_BYTES:
            return None

        length = int.from_bytes(self._buffer[:LENGTH_PREFIX_BYTES], "little")

        # Before anything is sliced or decoded. A corrupt or hostile prefix declaring
        # 4 GB must be an error now rather than an allocation first.
        if length > MAX_MESSAGE_BYTES:
            raise FramingError(f"declared length {length} exceeds the {MAX_MESSAGE_BYTES} byte limit")

        end = LENGTH_PREFIX_BYTES + length
        if len(self._buffer) < end:
            return None

        body = bytes(self._buffer[LENGTH_PREFIX_BYTES:end]).decode("utf-8", errors="replace")

        # Deleted from the front rather than tracked with an offset. The buffer holds at
        # most one partial message, so this copies kilobytes at worst -- and an offset
        # that is only compacted "sometimes" is how a long-lived reader grows without
        # bound.
        del self._buffer[:end]
        return body

    @property
    def pending(self) -> int:
        """Bytes held pending a complete frame.

        Non-zero between a partial read and the read that completes it. A value that
        only grows is a peer sending a prefix it never satisfies, which
        ``MAX_MESSAGE_BYTES`` bounds.
        """
        return len(self._buffer)

    def clear(self) -> None:
        self._buffer.clear()

File: engine/core/ipc/framing.cpp
#include "core/ipc/framing.h"

#include <algorithm>

namespace fc::ipc {

Result<std::vector<std::uint8_t>> encode_frame(std::string_view body) {
    if (body.size() > kMaxMessageBytes) {
        return FcError::IPC_MESSAGE_TOO_LARGE;
    }

    std::vector<std::uint8_t> frame;
    frame.reserve(kLengthPrefixBytes + body.size());

    // Little-endian by hand rather than by memcpy of a `std::uint32_t`: SPEC.md §15.1
    // specifies the byte order, and reading it off the host's happens to work on x64
    // and is not what the spec says.
    const auto length = static_cast<std::uint32_t>(body.size());
    frame.push_back(static_cast<std::uint8_t>(length & 0xFFU));
    frame.push_back(static_cast<std::uint8_t>((length >> 8U) & 0xFFU));
    frame.push_back(static_cast<std::uint8_t>((length >> 16U) & 0xFFU));
    frame.push_back(static_cast<std::uint8_t>((length >> 24U) & 0xFFU));

    frame.insert(frame.end(), body.begin(), body.end());
    return frame;
}

Result<DecodedFrame> decode_frame(std::span<const std::uint8_t> buffer) {
    if (buffer.size() < kLengthPrefixBytes) {
        return DecodedFrame{}; // not an error: the prefix has not fully arrived
    }

    const std::uint32_t length = static_cast<std::uint32_t>(buffer[0]) | (static_cast<std::uint32_t>(buffer[1]) << 8U) |
                                 (static_cast<std::uint32_t>(buffer[2]) << 16U) |
                                 (static_cast<std::uint32_t>(buffer[3]) << 24U);

    // Before any allocation, and before waiting for the body. A prefix declaring 4 GB
    // is either corruption or an attack, and in both cases the answer is to fail now
    // rather than to reserve first and find out afterwards.
    if (length > kMaxMessageBytes) {
        return FcError::IPC_MESSAGE_TOO_LARGE;
    }

    if (buffer.size() < kLengthPrefixBytes + length) {
        return DecodedFrame{}; // body still arriving
    }

    DecodedFrame decoded;
    const auto* start = reinterpret_cast<const char*>(buffer.data()) + kLengthPrefixBytes;
    decoded.body.assign(start, length);
    decoded.consumed = kLengthPrefixBytes + length;
    return decoded;
}

void FrameReader::append(std::span<const std::uint8_t> bytes) {
    buffer_.insert(buffer_.end(), bytes.begin(), bytes.end());
}

Result<std::optional<std::string>> FrameReader::next() {
    FC_TRY_ASSIGN(const DecodedFrame decoded, decode_frame(buffer_));
    if (decoded.consumed == 0) {
        return std::optional<std::string>{};
    }

    // Erase from the front rather than keeping an offset. The buffer holds at most one
    // partial message by construction, so this moves kilobytes at worst, and an offset
    // that is only compacted "sometimes" is how a long-lived reader grows without
    // bound -- which CLAUDE.md hard rule 5 forbids for exactly this reason.
    buffer_.erase(buffer_.begin(), buffer_.begin() + static_cast<std::ptrdiff_t>(decoded.consumed));
    return std::optional<std::string>{decoded.body};
}

} // namespace fc::ipc
